fms/testing/fms_hf_equivalence.py

```diff
@@ -0,0 +1,314 @@
+            v1 = v1[:v2_size]
+
+        elif size_diff < 0:
+            summary.extra_tokens += -size_diff
+            summary.sequences_with_extra_tokens.append(prompt_number)
+            v2 = v2[:v1_size]
+
+        summary.total_tokens += v1_size
+
+        diff = torch.sub(v1, v2)
+        num_diff = torch.count_nonzero(diff).item()
+        summary.different_tokens += num_diff
+        if num_diff > 0:
+            summary.sequences_with_different_tokens.append(prompt_number)
+
+    logging.info(f"diff summary = {json.dumps(summary.toJson(), indent=2)}")
+
+
+def main(argv):
+    logging.basicConfig(
+        level=logging.DEBUG,
+        format="%(asctime)s - %(levelname)s - %(message)s",
+        datefmt="%Y-%m-%d %H:%M:%S",
+    )
+
+    parser = argparse.ArgumentParser(
+        description="Utility functions for comparing outputs of two implementations of a model"
+    )
+    subparsers = parser.add_subparsers(dest="subcommand")
+
+    fms = subparsers.add_parser("run_fms_model", help="Run fms model")
+    fms.add_argument(
+        "-d", "--dtype", type=str, required=True, choices=["float16", "float32"]
+    )
+    fms.add_argument("-a", "--model_arch", type=str, required=True)
+    fms.add_argument("-v", "--model_variant", type=str, required=True)
+    fms.add_argument("-p", "--model_path", type=str, required=True)
+    fms.add_argument("-i", "--input_json_path", type=str, required=True)
+    fms.add_argument("-o", "--output_dir", type=str, required=True)
+
+    hf = subparsers.add_parser("run_hf_model", help="Run huggingface model")
+    hf.add_argument(
+        "-d", "--dtype", type=str, required=True, choices=["float16", "float32"]
+    )
+    hf.add_argument("-p", "--model_path", type=str, required=True)
+    hf.add_argument("-i", "--input_json_path", type=str, required=True)
+    hf.add_argument("-o", "--output_dir", type=str, required=True)
+
+    diff = subparsers.add_parser(
+        "compute_diff", help="Compute difference in two model outputs"
+    )
+    diff.add_argument("-f", "--generated_tokens_dir", type=str, required=True)
+
+    args = parser.parse_args(argv[1:])
+
+    if args.subcommand == "run_fms_model":
+        torch.set_default_dtype(dtype_mapping[args.dtype])
+        model, tokenizer = load_fms_model(
+            model_path=args.model_path,
+            model_arch=args.model_arch,
+            model_variant=args.model_variant,
+        )
+        generate_and_write_tokens(
+            model, tokenizer, args.input_json_path, args.output_dir, "fms"
+        )
+
+    elif args.subcommand == "run_hf_model":
+        torch.set_default_dtype(dtype_mapping[args.dtype])
+        model, tokenizer = load_hf_model(model_path=args.model_path)
+        generate_and_write_tokens(
+            model, tokenizer, args.input_json_path, args.output_dir, "hf"
+        )
+
+    elif args.subcommand == "compute_diff":
+        compute_diff(args.generated_tokens_dir)
```

Comment:
does this mean the user needs to run the script three times to get a diff?

Reply:
The idea is that the utility can be used to generate tokens under different settings in parallel. For instance, I mostly used to launch the runs for two models under float32, and float16 (so total 8 runs). If a particular run fails, it's easy to restart manually.

Once the corresponding outputs are available, the diff function is pretty lightweight and the user may want to run that afterward.

Having said that we can certainly enhance the utility to check if two GPUs are available, and run the models in parallel, otherwise serially. However, handling the failure scenarios (e.g., if one model fails, execute only the failed model on the next run) would require some work which would be worth investing in if over time we feel the usefulness of the utility.